domain/user/state/state.go

```diff
@@ -0,0 +1,248 @@
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// AddUser will add a new user to the database. If the user already exists
+// an error that satisfies usererrors.AlreadyExists will be returned. If the
+// users creator is set and does not exist then an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUser(ctx context.Context, domainUser domainuser.User) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	addUserQuery := "INSERT INTO user (uuid, name, display_name, created_by_uuid, created_at) VALUES (?, ?, ?, ?, ?)"
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		_, err := tx.ExecContext(ctx, addUserQuery,
+			domainUser.UUID,
+			domainUser.Name,
+			domainUser.DisplayName,
+			domainUser.CreatorUUID,
+			domainUser.CreatedAt,
+		)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+	if err != nil {
+		if databaseutils.IsErrConstraintUnique(err) {
+			return fmt.Errorf("user %q already exists: %w", domainUser.Name, err)
+		}
+		if databaseutils.IsErrConstraintForeignKey(err) {
+			return fmt.Errorf("user %q creator %q does not exist: %w", domainUser.Name, domainUser.CreatorUUID, err)
+		}
+		return fmt.Errorf("cannot insert user %q: %w", domainUser.Name, err)
+	}
+
+	return nil
+}
+
+// AddUserWithPasswordHash will add a new user to the database with the
+// provided password hash and salt. If the user already exists an error that
+// satisfies usererrors.AlreadyExists will be returned. If the users creator
+// does not exist or has been previously removed an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUserWithPasswordHash(ctx context.Context, domainUser domainuser.User, passwordHash string, salt []byte) error {
+	err := st.AddUser(ctx, domainUser)
+	if err != nil {
+		return fmt.Errorf("cannot add user %q: %w", domainUser.Name, err)
+	}
+
+	err = st.SetPasswordHash(ctx, domainUser.UUID, passwordHash, salt)
+	if err != nil {
+		return fmt.Errorf("cannot set password hash for user %q: %w", domainUser.Name, err)
+	}
+
+	return nil
+}
+
+// AddUserWithActivationKey will add a new user to the database with the
+// provided activation key. If the user already exists an error that
+// satisfies usererrors.AlreadyExists will be returned. if the users creator
+// does not exist or has been previously removed an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUserWithActivationKey(ctx context.Context, user domainuser.User, activationKey string) error {
+	return nil
+}
+
+// GetUser will retrieve the user specified by name from the database where
+// the user is active and has not been removed. If the user does not exist
+// or is deleted an error that satisfies usererrors.NotFound will be
+// returned.
+func (st *State) GetUser(ctx context.Context, userUUID string) (user.User, error) {
```

Comment:
I am still not seeing why we made such a big switch over to using uuid as a natural key for getting information in this domain.

When a user logs in into Juju we know their username and not their UUID. We should be working on usernames and not uuid's. Usernames are unique and are the single point of information that we work off across our entire code base.

UUID's are an accounting mechanism in state that don't have a need to bubble up for the moment.

Reply:
We will discuss this more tonight but at the moment it seems like we are bending over backwards to make UUID's work but we haven't considered the users of this state or service in the design and the sheer effort they are going to have to go to move across.